Fix `shard_module` to appropriately deal with sub process groups.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79264

`shard_module` API didn't work correctly with a sub-pg since
`dist.scatter` actually takes the global rank as input for `src`.

Fixing this by passing in the appropriate rank to `dist.scatter`

Differential Revision: [D37062766](https://our.internmc.facebook.com/intern/diff/D37062766/)

pritamdamania87 committed:
Fix `shard_module` to appropriately deal with sub process groups.

`shard_module` API didn't work correctly with a sub-pg since
`dist.scatter` actually takes the global rank as input for `src`.

Fixing this by passing in the appropriate rank to `dist.scatter`

Differential Revision: [D37062766](https://our.internmc.facebook.com/intern/diff/D37062766/)

[ghstack-poisoned]

pritamdamania87 committed:
Update on "Fix `shard_module` to appropriately deal with sub process groups."


`shard_module` API didn't work correctly with a sub-pg since
`dist.scatter` actually takes the global rank as input for `src`.

Fixing this by passing in the appropriate rank to `dist.scatter`

Differential Revision: [D37062766](https://our.internmc.facebook.com/intern/diff/D37062766/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py:
```diff
@@ -166,10 +167,15 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
         assert local_metadata is not None
 
         # Scatter the shards to all ranks in the pg
+        # scatter takes the global rank as ``src``
```
is the scatter behavior correct? I think scatter need to take local rank as `src` if it passes a different pg than the global pg.

pritamdamania87 replied:
I agree with you that scatter behavior doesn't seem right, but I'm not sure if we can change it without a BC breaking change.

pritamdamania87 replied:
Also, it looks like `gather`, `broadcast`, `send`, `recv` all have the same behavior. So this might probably need some deeper investigation.

wanchaol replied:
Yeah make sense, let's go ahead and merge it, fixing c10d can be a separate discussion.

pritamdamania87 commented:
@pytorchbot merge

pritamdamania87 committed:
Update on "Fix `shard_module` to appropriately deal with sub process groups."


`shard_module` API didn't work correctly with a sub-pg since
`dist.scatter` actually takes the global rank as input for `src`.

Fixing this by passing in the appropriate rank to `dist.scatter`

Differential Revision: [D37062766](https://our.internmc.facebook.com/intern/diff/D37062766/)

[ghstack-poisoned]

pritamdamania87 commented:
@pytorchbot merge
